src/Form/FormGroupContext.jsx

```diff
@@ -1,61 +1,82 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
 import PropTypes from 'prop-types';
+import classNames from 'classnames';
 import { newId } from '../utils';
-import useToggle from '../hooks/useToggle';
-import {
-  callAllHandlers,
-  useIdList,
-  mergeAttributeValues,
-  omitUndefinedProperties,
-} from './fieldUtils';
+import { useIdList, omitUndefinedProperties } from './fieldUtils';
 import { FORM_CONTROL_SIZES } from './constants';
 
-const FormGroupContext = React.createContext();
+const identityFn = props => props;
+const noop = () => {};
 
-const defaultFormGroupContextValue = {
-  getNewDescriptorId: () => {}, // noop by default
-};
+const FormGroupContext = React.createContext({
+  getControlProps: identityFn,
+  useSetControlIsGroupEffect: noop,
+  getLabelProps: identityFn,
+  getDescriptorProps: identityFn,
+});
+
+const useFormGroupContext = () => React.useContext(FormGroupContext);
 
-const useFormGroupContext = (props = {}) => {
-  const cleanedProps = omitUndefinedProperties(props);
-  const contextValue = React.useContext(FormGroupContext) || {};
-  const mergedValues = {
-    ...defaultFormGroupContextValue,
-    ...contextValue,
-    ...cleanedProps,
-    'aria-describedby': mergeAttributeValues(
-      contextValue['aria-describedby'],
-      props['aria-describedby'],
-    ),
-    onFocus: callAllHandlers(contextValue.onFocus, props.onFocus),
-    onBlur: callAllHandlers(contextValue.onBlur, props.onBlur),
+const useStateEffect = (initialState) => {
+  const [state, setState] = useState(initialState);
+  const useSetStateEffect = (newState) => {
+    useEffect(() => setState(newState), [newState]);
   };
-  return mergedValues;
+  return [state, useSetStateEffect];
 };
 
 const FormGroupContextProvider = ({
   children,
   controlId,
   isInvalid,
   isValid,
-  onFocus,
-  onBlur,
   size,
 }) => {
   const resolvedId = React.useMemo(() => controlId || newId('form-field'), [controlId]);
-  const [hasFocus, setHasFocusTrue, setHasFocusFalse] = useToggle(false);
-  const [describedByIds, getNewDescriptorId, removeDescriptorId] = useIdList(resolvedId);
+  const [describedByIds, useRegisteredDescriptorId] = useIdList(resolvedId);
+  const [labelledByIds, useRegisteredLabellerId] = useIdList(resolvedId);
+  const [controlIsGroup, useSetControlIsGroupEffect] = useStateEffect(false);
```

Comment:
naming suggestion/nit: `isControlGroup` vs. `controlIsGroup` (bit more indicative of a boolean).

Reply:
👍